storage/mysql/mysql_test.go

```diff
@@ -28,14 +63,253 @@ func TestMain(m *testing.M) {
+	}
+
+	db, err := sql.Open("mysql", *mysqlURI+"?multiStatements=true")
+	if err != nil {
+		klog.Fatalf("Failed to connect to DB: %v", err)
+	}
+	defer func() {
+		if err := db.Close(); err != nil {
+			klog.Warningf("Failed to close db: %v", err)
+		}
+	}()
+
+	if _, err := db.ExecContext(ctx, dropTablesSQL); err != nil {
+		klog.Fatalf("Failed to drop all tables: %v", err)
+	}
+
+	if _, err := db.ExecContext(ctx, string(rawSchema)); err != nil {
+		klog.Fatalf("Failed to execute init database schema: %v", err)
+	}
+}
+
+func TestNew(t *testing.T) {
+	ctx := context.Background()
+
+	for _, test := range []struct {
+		name    string
+		opts    []func(*tessera.StorageOptions)
+		wantErr bool
+	}{
+		{
+			name:    "no tessera.StorageOption",
+			opts:    nil,
+			wantErr: true,
+		},
+		{
+			name: "standard tessera.WithCheckpointSignerVerifier",
+			opts: []func(*tessera.StorageOptions){
+				tessera.WithCheckpointSignerVerifier(noteSigner, noteVerifier),
+			},
+			wantErr: false,
+		},
+		{
+			name: "all tessera.StorageOption",
+			opts: []func(*tessera.StorageOptions){
+				tessera.WithCheckpointSignerVerifier(noteSigner, noteVerifier),
+				tessera.WithBatching(1, 1*time.Second),
+				tessera.WithPushback(10),
+			},
+			wantErr: false,
+		},
+	} {
+		t.Run(test.name, func(t *testing.T) {
+			_, err := mysql.New(ctx, testDB, test.opts...)
+			gotErr := err != nil
+			if gotErr != test.wantErr {
+				t.Errorf("got err %v want %v", gotErr, test.wantErr)
+			}
+		})
+	}
+}
+
+func TestReadTile(t *testing.T) {
+	ctx := context.Background()
+	s := newTestMySQLStorage(t, ctx)
+
+	for _, test := range []struct {
+		name                string
+		level, index, width uint64
+		wantErr             bool
+	}{
+		{
+			name:  "0/0/0",
+			level: 0, index: 0, width: 0,
+			wantErr: false,
+		},
+		{
+			name:  "123/456/789",
+			level: 123, index: 456, width: 789,
+			wantErr: false,
+		},
+	} {
+		t.Run(test.name, func(t *testing.T) {
+			_, err := s.ReadTile(ctx, test.level, test.index, test.width)
+			gotErr := err != nil
+			if gotErr != test.wantErr {
+				t.Errorf("got err %v want %v", gotErr, test.wantErr)
+			}
+		})
+	}
+}
+
+func TestReadEntryBundle(t *testing.T) {
+	ctx := context.Background()
+	s := newTestMySQLStorage(t, ctx)
+
+	for _, test := range []struct {
+		name    string
+		index   uint64
+		wantErr bool
+	}{
+		{
+			name:    "0",
+			index:   0,
+			wantErr: false,
+		},
+		{
+			name:    "123456789",
+			index:   123456789,
+			wantErr: false,
+		},
+	} {
+		t.Run(test.name, func(t *testing.T) {
+			_, err := s.ReadEntryBundle(ctx, test.index)
+			gotErr := err != nil
+			if gotErr != test.wantErr {
+				t.Errorf("got err %v want %v", gotErr, test.wantErr)
+			}
+		})
+	}
+}
+
+func TestAdd(t *testing.T) {
+	ctx := context.Background()
+	s := newTestMySQLStorage(t, ctx)
+
+	for _, test := range []struct {
+		name      string
+		entry     []byte
+		wantIndex uint64
+		wantErr   bool
```

Comment:
This is always `false`, could it be removed?

Reply:
Done.